### PR TITLE
[Wallet] refactor wallet/init interaction

### PR DESCRIPTION
**Pull request by jonasschnelli**
This is another step to move the wallet code from `init.cpp` to `wallet.cpp` while keeping the behavior identical. Behavior simplification can be done once all code is available in `wallet.cpp`.

This is also a basic step into the direction of cloning the wallet, add BIP32 & multi-wallet in the second wallet implementation without introducing bugs in the current wallet implementation.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Concept ACK. Please remove the parameter interaction code from init; it seems you forgot to do this.

**Comment by jonasschnelli** [comment]
@MarcoFalke: thanks! Fixed.

**Comment by paveljanik** [comment]
New warning:
```
wallet/wallet.cpp:3222:1: warning: control may reach end of non-void function [-Wreturn-type]
}
^
1 warning generated.
```

**Comment by jtimon** [comment]
Could ```CWallet::ParameterInteraction()``` take a ```const std::map<std::string, std::string>& mapArgs``` as parameter instead of using the mapArgs global from util.h  like in https://github.com/jtimon/bitcoin/commit/3d03f15126b79bd9f22ccd27f1d8ca43d9e16bf5 ?

In fact, I believe I got the idea from one of your older versions of this (and then suggested some improvements to std::vector<std::pair<std::string, std::string> > CStandardPolicy::GetOptionsHelp() that would allow him to create generic forms in qt that automatically extend themselves when new options are added).

Concept ACK

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -875,8 +877,11 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
     /* Returns the wallets help message */
     static std::string GetWalletHelpString(bool showDebug);
 
-    /* initializes the wallet, returns a new CWallet instance or a null pointer in case of an error */
-    static CWallet* InitLoadWallet(bool fDisableWallet, const std::string& strWalletFile, std::string& warningString, std::string& errorString);
+    /* Initializes the wallet, returns a new CWallet instance or a null pointer in case of an error */
+    static bool InitLoadWallet();
```
Comment (jtimon) [inline src/wallet/wallet.h]: Why static instead of calling it from the pwalletMain global (which could be moved to wallet)?
Or are we turning the global into a singleton?

Comment (jonasschnelli) [inline src/wallet/wallet.h]: > Why static instead of calling it from the pwalletMain global (which could be moved to wallet)?
> Or are we turning the global into a singleton?

Right. This is the direction where I'm heading to. But I work in very tiny steps (as you can see). For now it just moves "static code" to a static function in CWallet. Later it should be within the `pwalletMain` context.

File: src/init.cpp [hunk 2]
```diff
@@ -946,56 +943,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
-            return InitError(strprintf(_("Invalid amount for -maxtxfee=<amount>: '%s' (must be at least the minrelay fee of %s to prevent stuck transactions)"),
-                                       mapArgs["-maxtxfee"], ::minRelayTxFee.ToString()));
-        }
-    }
-    nTxConfirmTarget = GetArg("-txconfirmtarget", DEFAULT_TX_CONFIRM_TARGET);
-    bSpendZeroConfChange = GetBoolArg("-spendzeroconfchange", DEFAULT_SPEND_ZEROCONF_CHANGE);
-    fSendFreeTransactions = GetBoolArg("-sendfreetransactions", DEFAULT_SEND_FREE_TRANSACTIONS);
-
-    std::string strWalletFile = GetArg("-wallet", DEFAULT_WALLET_DAT);
+    CWallet::ParameterInteraction();
```
Comment (MarcoFalke) [inline src/init.cpp]: You can't drop the return value here.

Comment (jonasschnelli) [inline src/init.cpp]: Right. Fixed.

**Comment by MarcoFalke** [comment]
You forgot to remove `AmountErrMsg` from init as well. Maybe it would help to create two separate commits or even pulls: One addressing the move-only and the other the refactoring Warning/Error messages?

I really want to get this merged quick but I am concerned that reviewers are turned off by the diff containing both moves and refactors.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -364,8 +365,34 @@ void CWallet::Flush(bool shutdown)
...
+    uiInterface.ThreadSafeMessageBox(str, "", CClientUIInterface::MSG_ERROR);
+    return false;
+}
+
+bool static UIWarning(const std::string &str)
+{
+    uiInterface.ThreadSafeMessageBox(str, "", CClientUIInterface::MSG_WARNING);
+    return true;
+}
+
```
Comment (mrbandrews) [inline src/wallet/wallet.cpp]: I think this should be void, as the return value is not used as far as I can tell.

**Comment by mrbandrews** [comment]
AmountErrMsg still used once in init.cpp at line 937 of the revised file.  So it's defined as a static function in both wallet.cpp and init.cpp, which seems to work fine, although I don't know whether it's the best way to do it (maybe it is)

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: It is mirroring `bool static InitWarning()`, so probably fine. Still, the code duplication could be avoided somehow.

Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: I have thought about that and if we want to decouple the wallet, we probably need to accept short term duplications of some code. In that case, the factored out shared code is `uiInterface.ThreadSafeMessageBox`. A duplicated entry point only used by CWallet should be okay.

But I agree with @mrbandrews: the return code should be switched to `void` (for `UIWarning`, keep it for `UIError` for now).
Will fix that.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3150,7 +3163,64 @@ CWallet* CWallet::InitLoadWallet(bool fDisableWallet, const std::string& strWall
...
+            return UIError(strprintf(_("Invalid amount for -maxtxfee=<amount>: '%s' (must be at least the minrelay fee of %s to prevent stuck transactions)"),
+                                       mapArgs["-maxtxfee"], ::minRelayTxFee.ToString()));
+        }
+    }
+    nTxConfirmTarget = GetArg("-txconfirmtarget", DEFAULT_TX_CONFIRM_TARGET);
+    bSpendZeroConfChange = GetBoolArg("-spendzeroconfchange", DEFAULT_SPEND_ZEROCONF_CHANGE);
+    fSendFreeTransactions = GetBoolArg("-sendfreetransactions", DEFAULT_SEND_FREE_TRANSACTIONS);
+
+    std::string strWalletFile = GetArg("-wallet", DEFAULT_WALLET_DAT);
+
```
Comment (mrbandrews) [inline src/wallet/wallet.cpp]: I think this line can be removed... the wallet file name is read at the beginning of InitLoadWallet.

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -2992,8 +3017,10 @@ std::string CWallet::GetWalletHelpString(bool showDebug)
     return strUsage;
 }
 
-CWallet* CWallet::InitLoadWallet(bool fDisableWallet, const std::string& strWalletFile, std::string& warningString, std::string& errorString)
+bool CWallet::InitLoadWallet()
 {
+    std::string strWalletFile = GetArg("-wallet", DEFAULT_WALLET_DAT);
+
```
Comment (mrbandrews) [inline src/wallet/wallet.cpp]: Nit:  this local variable has the same name as the class member variable (I guess it is not a problem here since the compiler knows it can't access the class member from a static method?)  Not a big deal but maybe want to give it a different name.

**Comment by mrbandrews** [comment]
Other than a couple of nits, this looks good to me.  Code review ACK, concept ACK, and I tested it with an unpruned node, pruned node with wallet, and a pruned node with an outdated wallet - meaning it should suggest a reindex (qt) or shut down (bitcoind) - with no problems.  So it's an ACK from me.

**Comment by jonasschnelli** [comment]
Fixed nits.
Should be ready for merge.

**Comment by morcos** [comment]
@jonasschnelli I don't necessarily object to this but it seems like if you're eventually going to change these from static functions to methods called on pWalletMain, that you might as well make all those changes now instead of having to review twice.
If you feel like its cleaner to break it up, perhaps it could be 2 commits in the same PR, but why not move closer to where you're trying to go.

**Comment by jonasschnelli** [comment]
Okay. Fair enough.
Added two commits.
First added commit will refactor the static methods to take them into the CWallet instance context. This looks much cleaner now.

Second commit is a cosmetic BDB version logprint change.

A migration from the global `pwalletMain` (called from everywhere) to a singleton is out-of-scope for now.

Thanks for the reviews!

**Comment by jonasschnelli** [comment]
Rejecting @morcos proposal to integrate the static-to-context change. Rolling back to original commit. This would get to big and not enough reviews. Further refactor can be done after this PR.

**Comment by MarcoFalke** [comment]
Great, thanks for addressing the other nits.

utACK 25340b7

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -3150,7 +3162,62 @@ CWallet* CWallet::InitLoadWallet(bool fDisableWallet, const std::string& strWall
     }
     walletInstance->SetBroadcastTransactions(GetBoolArg("-walletbroadcast", DEFAULT_WALLETBROADCAST));
 
-    return walletInstance;
+    pwalletMain = walletInstance;
+    return true;
+}
+
+bool CWallet::ParameterInteraction()
```
Comment (jtimon) [inline src/wallet/wallet.cpp]: Why not pass mapArgs as parameter here?

Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: The goal was it to keep it at a tiny scope (almost moveonly). I would strongly advise to keep this PR as it is and refactor later.

Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Appears out of scope for this pull.

Comment (jtimon) [inline src/wallet/wallet.cpp]: Never mind, it won't make much sense until you can explicitly pass it to GetArg(), like in https://github.com/bitcoin/bitcoin/pull/6423/commits/3d03f15126b79bd9f22ccd27f1d8ca43d9e16bf5#diff-b7702a084eb00fb47f9800fd68271951R315

**Comment by jtimon** [comment]
utACK, all my nits can be solved later.

**Comment by laanwj** [comment]
utACK 

> A migration from the global pwalletMain (called from everywhere) to a singleton is out-of-scope for now.

Seems like a useless intermediate step to me.
Singletons are just nicely dressed-up globals.
But we'd like to *get rid* of a single, global wallet, in any form. Just pass the thing around when you need it.
